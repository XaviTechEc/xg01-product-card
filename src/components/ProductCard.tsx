import React, { createContext } from 'react';

import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import {
	InitialValues,
	IOnChangeArgs,
	IProduct,
	IProductContextProps,
	ProductCardHandlers
} from '../interfaces/Product';

// import { ProductTitle, ProductButtons, ProductImage } from './';

export const ProductContext = createContext({} as IProductContextProps);
export const { Provider } = ProductContext;

export type ProductCardProps = {
	product: IProduct;
	// children?: React.ReactElement | React.ReactElement[];
	children: (args: ProductCardHandlers) => JSX.Element;
	className?: string;
	style?: React.CSSProperties;
	onChange?: (args: IOnChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
};

export const ProductCard = ({
	children,
	product,
	className,
	style,
	onChange,
	value,
	initialValues
}: ProductCardProps) => {
	const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues
		});

	//* Values available in the provider
	const values = { counter, increaseBy, product, maxCount };

	return (
		<Provider value={values}>
			<div
				className={`${styles.productCard} ${className}`}
				style={style}
			>
				{
					//* === FUNTION CHILD ===
					children({
						count: counter,
						isMaxCountReached,
						maxCount,
						product,
						increaseBy,
						reset
					})
				}
			</div>
		</Provider>
	);
};

// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;
