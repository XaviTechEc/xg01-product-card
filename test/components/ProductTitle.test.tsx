import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard, { ProductTitle } from '../../src/components';

import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('Must show the component with a custom title', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom Product" className="custom-class" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Must show title with the name of a product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => {
          return <ProductTitle />;
        }}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
