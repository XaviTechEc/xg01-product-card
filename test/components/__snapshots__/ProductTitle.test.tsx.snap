// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Must show the component with a custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Product
</span>
`;

exports[`ProductTitle Must show title with the name of a product 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Product Card NO Img
  </span>
</div>
`;
